panic: close of closed channel

when trying to run locally against a cardano-node running in a docker container

```
panic: close of closed channel

goroutine 6119927 [running]:
github.com/blinklabs-io/gouroboros.(*Connection).shutdown(0xc0000ec420)
	/home/chris/go/pkg/mod/github.com/blinklabs-io/gouroboros@v0.147.0/connection.go:266 +0x54
github.com/blinklabs-io/gouroboros.(*Connection).setupConnection.func1()
	/home/chris/go/pkg/mod/github.com/blinklabs-io/gouroboros@v0.147.0/connection.go:283 +0x2f
created by github.com/blinklabs-io/gouroboros.(*Connection).setupConnection in goroutine 6131411
	/home/chris/go/pkg/mod/github.com/blinklabs-io/gouroboros@v0.147.0/connection.go:281 +0xb2

```

--- a/main.go
+++ b/main.go
@@ -230,14 +230,6 @@ func GetConnection(errorChan chan error) (*ouroboros.Connection, error) {
 	// it and cause a "send on closed channel" panic on the next connection.
 	// Instead give each connection its own channel and relay errors out.
 	connErrorChan := make(chan error, 10)
-	go func() {
-		for err := range connErrorChan {
-			select {
-			case errorChan <- err:
-			default:
-			}
-		}
-	}()
 	oConn, err := ouroboros.NewConnection(
 		ouroboros.WithNetworkMagic(uint32(cfg.Node.NetworkMagic)),
 		ouroboros.WithErrorChan(connErrorChan),
@@ -293,6 +285,14 @@ func GetConnection(errorChan chan error) (*ouroboros.Connection, error) {
 			return nil, errors.New("specify either the UNIX socket path or the address/port")
 		}
 		slog.Info("Successfully connected to node")
+		go func() {
+			for err := range connErrorChan {
+				select {
+				case errorChan <- err:
+				default:
+				}
+			}
+		}()
 		return oConn, nil
 	}
 	return nil, fmt.Errorf(

--- a/main.go
+++ b/main.go
@@ -225,9 +225,22 @@ func (c *Config) populateNetworkMagic() error {
 
 func GetConnection(errorChan chan error) (*ouroboros.Connection, error) {
 	cfg := GetConfig()
+	// gouroboros closes the error channel it's given during shutdown, so we
+	// must never hand it the shared errorChan directly — doing so would close
+	// it and cause a "send on closed channel" panic on the next connection.
+	// Instead give each connection its own channel and relay errors out.
+	connErrorChan := make(chan error, 10)
+	go func() {
+		for err := range connErrorChan {
+			select {
+			case errorChan <- err:
+			default:
+			}
+		}
+	}()
 	oConn, err := ouroboros.NewConnection(
 		ouroboros.WithNetworkMagic(uint32(cfg.Node.NetworkMagic)),
-		ouroboros.WithErrorChan(errorChan),
+		ouroboros.WithErrorChan(connErrorChan),
 		ouroboros.WithNodeToNode(false),
 		ouroboros.WithKeepAlive(true),
 	)
@@ -559,6 +572,7 @@ func initializeData(errorChan chan error) {
 			GetSizes(oConn),
 			GetTransactions(oConn),
 		))
+		oConn.Close()
 	}
 }
 
@@ -673,6 +687,7 @@ func startRefreshLoop(cfg *Config, errorChan chan error) {
 						GetSizes(oConn),
 						GetTransactions(oConn),
 					)
+					oConn.Close()
 					if tmpText != "" && tmpText != content {
 						content = tmpText
 						text.Clear()

--- a/main.go
+++ b/main.go
@@ -287,10 +287,7 @@ func GetConnection(errorChan chan error) (*ouroboros.Connection, error) {
 		slog.Info("Successfully connected to node")
 		go func() {
 			for err := range connErrorChan {
-				select {
-				case errorChan <- err:
-				default:
-				}
+				errorChan <- err
 			}
 		}()
 		return oConn, nil

--- a/main_test.go
+++ b/main_test.go
@@ -205,7 +205,7 @@ func TestTogglePaused(t *testing.T) {
 }
 
 func TestErrorChanIsolation(t *testing.T) {
-	// Regression test for issue #287: panic: send on closed channel.
+	// Regression test for issue panic: send on closed channel.
 	//
 	// gouroboros closes whichever error channel it is given during connection
 	// shutdown. Passing the same shared errorChan to every connection means

--- a/main_test.go
+++ b/main_test.go
@@ -224,10 +224,7 @@ func TestErrorChanIsolation(t *testing.T) {
 		go func() {
 			defer close(done)
 			for err := range connErrChan {
-				select {
-				case sharedErrChan <- err:
-				default:
-				}
+				sharedErrChan <- err
 			}
 		}()
 		connErrChan <- fmt.Errorf("error from connection %d", id)

--- a/main_test.go
+++ b/main_test.go
@@ -204,6 +204,45 @@ func TestTogglePaused(t *testing.T) {
 	}
 }
 
+func TestErrorChanIsolation(t *testing.T) {
+	// Regression test for issue #287: panic: send on closed channel.
+	//
+	// gouroboros closes whichever error channel it is given during connection
+	// shutdown. Passing the same shared errorChan to every connection means
+	// the first shutdown closes it, and the second connection panics when it
+	// tries to send on the now-closed channel.
+	//
+	// The fix gives each connection its own connErrChan and relays errors into
+	// the shared channel. This test verifies that pattern: two simulated
+	// connection lifecycles must not panic and must relay their errors.
+
+	sharedErrChan := make(chan error, 10)
+
+	simulateConnection := func(id int) {
+		connErrChan := make(chan error, 10)
+		done := make(chan struct{})
+		go func() {
+			defer close(done)
+			for err := range connErrChan {
+				select {
+				case sharedErrChan <- err:
+				default:
+				}
+			}
+		}()
+		connErrChan <- fmt.Errorf("error from connection %d", id)
+		close(connErrChan) // gouroboros closes this on shutdown
+		<-done             // wait for relay goroutine to finish
+	}
+
+	simulateConnection(1)
+	simulateConnection(2) // would panic before the fix
+
+	if len(sharedErrChan) != 2 {
+		t.Errorf("expected 2 relayed errors, got %d", len(sharedErrChan))
+	}
+}
+
 func TestLogBuffer_Write(t *testing.T) {
 	lb := &LogBuffer{maxLines: 3}
 

